File: Interviews/GFG/random-problems/Easy/Longest-Perfect-Piece/Longest-Perfect-Piece.py
from sortedcontainers import SortedList
class Solution:
    def longestPerfectPiece(self, arr, N):
        subarray, ans, j =  SortedList(), 0, 0
        for idx, i in enumerate(arr):
            subarray.add(i)
            while subarray[-1] - subarray[0] > 1:
                subarray.discard(arr[j])
                j += 1
            ans = max(ans, idx - j + 1)
        
        return ans

File: Interviews/GFG/Daily-Problems/2022/November/Day-15/Longest-Perfect-Piece.cpp
class Solution {
   public:
    int longestPerfectPiece(int arr[], int N) {
        multiset<int> ms;
        int ans = 0, j = 0;
        for (int i = 0; i < N; i++) {
            ms.insert(arr[i]);
            while (*ms.rbegin() - (*ms.begin()) > 1) {
                ms.erase(arr[j]);
                j += 1;
            }
            ans = max(ans, i - j + 1);
        }

        return ans;
    }
};